contrib/opencensus-ext-azure/opencensus/ext/azure/common/transport.py

```diff
@@ -0,0 +1,133 @@
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import json
+import logging
+import requests
+
+from opencensus.trace import execution_context
+
+logger = logging.getLogger(__name__)
+
+
+class TransportMixin(object):
+    def _transmit_from_storage(self):
+        for blob in self.storage.gets():
+            if blob.lease(self.options.timeout + 5):
+                envelopes = blob.get()  # TODO: handle error
+                result = self._transmit(envelopes)
+                if result > 0:
+                    blob.lease(result)
+                else:
+                    blob.delete(silent=True)
+
+    def _transmit(self, envelopes):
+        """
+        Transmit the data envelopes to the ingestion service.
+        Return a negative value for partial success or non-retryable failure.
+        Return 0 if all envelopes have been successfully ingested.
+        Return the next retry time in seconds for retryable failure.
+        This function should never throw exception.
+        """
+        # TODO: prevent requests being tracked
+        blacklist_hostnames = execution_context.get_opencensus_attr(
+            'blacklist_hostnames',
+        )
+        execution_context.set_opencensus_attr(
+            'blacklist_hostnames',
+            ['dc.services.visualstudio.com'],
+        )
+        try:
+            response = requests.post(
+                url=self.options.endpoint,
+                data=json.dumps(envelopes),
+                headers={
+                    'Accept': 'application/json',
+                    'Content-Type': 'application/json; charset=utf-8',
+                },
+                timeout=self.options.timeout,
+            )
+        except Exception as ex:  # TODO: consider RequestException
+            logger.warning('Transient client side error %s.', ex)
+            # client side error (retryable)
+            return self.options.minimum_retry_interval
+        finally:
+            execution_context.set_opencensus_attr(
+                'blacklist_hostnames',
+                blacklist_hostnames,
+            )
+        text = 'N/A'
+        data = None
+        try:
+            text = response.text
+        except Exception as ex:
+            logger.warning('Error while reading response body %s.', ex)
+        else:
+            try:
+                data = json.loads(text)
+            except Exception:
+                pass
+        if response.status_code == 200:
+            logger.info('Transmission succeeded: %s.', text)
```

Comment:
Nit: I feel there's a bit too much logs in this class. Maybe consider reducing logs or using traces with a low sampling rate.

Reply:
This is at info level, which is intended for debugging only and off by default.
The default log level is WARNING.